### PR TITLE
Load basic user data to redux async, initialized in header

### PR DESCRIPTION
**Pull request by maureensturgeon**
This is a portion of the work required to make the teacher panel load on cached pages (https://github.com/code-dot-org/code-dot-org/pull/43126).

Changes include:
- Creating a new endpoint to query data about the current user
- The header dispatches the query to get the current user and set the result to `currentUserRedux`, this request is made in the header to ensure that this data is available across all pages that include the header
- Removes a variety of actions to dispatch setting current user information for different pages since it's now happening in the header

## Testing story
Added unit tests and verified that current user data is being set on various pages.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by jamescodeorg** [comment]
Looks good to me, but I'm hoping someone who understand how the header interacts with other pages can chime in.  Maybe Brendan would be a good person to add to this PR?

File: apps/src/templates/currentUserRedux.js [hunk 1]
```diff
@@ -60,12 +53,40 @@ export default function currentUser(state = initialState, action) {
...
+      userName: username,
+      userType: user_type
+    };
+  }
+
   return state;
 }
+
+export const asyncLoadUserData = () => dispatch => {
+  $.ajax('/api/v1/users/current_user')
```
Comment (tim-dot-org) [inline apps/src/templates/currentUserRedux.js]: If we're adding new ajax calls, could we use fetch here? we've got a polyfill that makes it work in all browsers

**Review by tim-dot-org — COMMENTED** [review]
the redux side looks fine!

File: dashboard/app/controllers/api/v1/users_controller.rb [hunk 2]
```diff
@@ -12,6 +13,22 @@ def load_user
...
   end
 
+  # GET /api/v1/users/current_user
+  def get_current_user
+    if current_user
+      render json: {
+        id: current_user.id,
+        username: current_user.username,
+        user_type: current_user.user_type,
+        is_signed_in: true
```
Comment (jamescodeorg) [inline dashboard/app/controllers/api/v1/users_controller.rb:23]: Would returning verified teacher and setting it in redux be a good idea?  Or is that too far from "basic" user information?

Also, while we're at it, do we want to handle display name here?

Comment (maureensturgeon) [reply dashboard/app/controllers/api/v1/users_controller.rb:23]: I like the idea of setting verified teacher too, what stopped me from doing that is that we have a separate verifiedTeacher redux..

**Comment by breville** [comment]
I like this idea, but it's a great question whether we could load it somewhere more appropriate than header-related code.  

Without context, I presume there's a certain class of data we'd like to have available for things that appear anywhere on the screen, be it in header, in a lab, or in a panel that pops over the screen (like the teacher panel).  

If we one day improve these pages to be rendered by React from top-to-bottom, then we should do the appropriate thing and scope logic to the bounding boxes on the screen, meaning we'd have an outermost container, and then a component for the header, a component for the level, etc.  

Putting this code in `header.js` feels just a little bit like it's the most convenient, but not the most architecturally correct.  The question is whether the more correct place currently exists.

cc @madelynkasula who has been thinking about these questions too.

For me, no big objections, and it's such a small piece of code that it can easily move around later, but I remain curious whether it could live somewhere just slightly more general than something designed to manage only the header area.

File: dashboard/config/routes.rb [hunk 3]
```diff
@@ -751,6 +751,7 @@
       get 'users/:user_id/using_text_mode', to: 'users#get_using_text_mode'
       get 'users/:user_id/display_theme', to: 'users#get_display_theme'
       get 'users/:user_id/contact_details', to: 'users#get_contact_details'
+      get 'users/current_user', to: 'users#get_current_user'
```
Comment (maddiedierker) [inline dashboard/config/routes.rb]: nit: would it be more readable/RESTful if this route/method were:

```ruby
get 'users/current', to: 'users#current'
```

**Review by maddiedierker — COMMENTED** [review]
i think your approach makes sense here. i'd love for us to use [commonReducers.js](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/redux/commonReducers.js) for this purpose (you'd think these reducers would be registered on every page load...), but i think that would be a much larger change/refactor

**Comment by maddiedierker** [comment]
just saw @breville's comment. i 100% agree with brendan; with our current architecture, i can't think of a different place to put this code since we have no single entrypoint for our apps code. this code is easy enough to move, and this sort of problem is top-of-mind when we think about changing our frontend architecture.

**Review by maureensturgeon — COMMENTED** [review]
Thanks everyone for the feedback! I agree if we had a wrapper that'd be a better place for this. I'm going to move forward with the solution I have here for now

File: dashboard/app/controllers/api/v1/users_controller.rb (hunk 2 above)
Comment (maureensturgeon) [reply dashboard/app/controllers/api/v1/users_controller.rb:23]: Circling back: James and I had a discussion about this, I like the idea of also setting verified teacher globally, I'm just not going to include that as part of these changes

**Review by jamescodeorg — APPROVED** [review]
For my education, if I want to use these user values, how can I write my code so that it only executes after these values have been retrieved?

**Comment by maureensturgeon** [comment]
> For my education, if I want to use these user values, how can I write my code so that it only executes after these values have been retrieved?

You can connect to redux and check that currentUser.signInState is not unknown

**Pull request merged**